src/ngMessages/messages.js

```diff
@@ -654,16 +654,17 @@ function ngMessageDirectiveFactory() {
           assignRecords(staticExp);
         }
 
-        var currentElement, messageCtrl;
+        var currentElement, currentScope, messageCtrl;
         ngMessagesCtrl.register(commentNode, messageCtrl = {
           test: function(name) {
             return contains(records, name);
           },
           attach: function() {
             if (!currentElement) {
-              $transclude(scope, function(elm) {
+              $transclude(function(elm, newScope) {
                 $animate.enter(elm, null, element);
                 currentElement = elm;
+                currentScope = newScope;
```

Comment:
Another alternative, is to bind to the element's `$destroy` event and clean up from there:

```js
elm.on('$destroy', function () { newScope.$destroy(); });
```


It has the benefits that you don't need to keep track of the `currentScope` and you are "safe" in case the element is somehow removed without going through the `detach()` method (although that's probably not possible, but better safe than sorry).